csrc/multidevice/utils.cpp

```diff
@@ -400,47 +420,67 @@ bool haveDifferentShardings(
        consumer->getMaybeRootDomain().size()) *
       2);
 
-  auto create_index = [&](IterDomain* id) {
+  auto create_index = [&](IterDomain* id, bool mapped) {
     auto* index = IrBuilder::create<Val>(DataType::Index);
-    NVF_ERROR(id_to_index.emplace(id, std::make_pair(index, false)).second);
+    NVF_ERROR(id_to_index.emplace(id, std::make_pair(index, mapped)).second);
     assumptions.push_back(
         SimplifyingIrBuilder::leExpr(fusion->zeroVal(), index));
     assumptions.push_back(SimplifyingIrBuilder::ltExpr(index, id->extent()));
   };
 
-  for (IterDomain* p_id : producer->getLogicalDomain()) {
```

Comment:
Previously, we create loop indices for all producer IDs and consumer IDs, regardless of whether they are used by DIDs. With this PR, we create 6x fewer exprs and vals.